library/std/src/io/error/repr_bitpacked.rs

```diff
@@ -0,0 +1,346 @@
+//! This is a densely packed error representation which is used on targets with
+//! 64-bit pointers.
+//!
+//! (Note that `bitpacked` vs `unpacked` here has no relationship to
+//! `#[repr(packed)]`, it just refers to attempting to use any available
+//! bits in a more clever manner than `rustc`'s default layout algorithm would).
+//!
+//! Conceptually, it stores the same information as the "unpacked" equivalent we
+//! use on other targets: `repr_unpacked::Repr` (see repr_unpacked.rs), however
+//! it packs it into a 64bit non-zero value.
+//!
+//! This optimization not only allows `io::Error` to occupy a single pointer,
+//! but improves `io::Result` as well, especially for situations like
+//! `Result<()>` (which is now 64 bits) or `Result<u64>` (which i), which are
+//! quite common.
+//!
+//! # Layout
+//! Tagged values are 64 bits, with the 2 least significant bits used for the
+//! tag. This means there are there are 4 "variants":
+//!
+//! - **Tag 0b00**: The first variant is equivalent to
+//!   `ErrorData::SimpleMessage`, and holds a `&'static SimpleMessage` directly.
+//!
+//!   `SimpleMessage` has an alignment >= 4 (which is requested with
+//!   `#[repr(align)]` and checked statically at the bottom of this file), which
+//!   means every `&'static SimpleMessage` should have the both tag bits as 0,
+//!   meaning its tagged and untagged representation are equivalent.
+//!
+//!   This means we can skip tagging it, which is necessary as this variant can
+//!   be constructed from a `const fn`, which probably cannot tag pointers (or
+//!   at least it would be difficult.
+//!
+//! - **Tag 0b01**: The other pointer variant holds the data for
+//!   `ErrorData::Custom` and the remaining 62 bits are used to store a
+//!   `Box<Custom>`. `Custom` also has alignment >= 4, so the bottom two bits
+//!   are free to use for the tag.
+//!
+//!   The only important thing to note is that `ptr::add` and `ptr::sub` are
+//!   used to tag the pointer, rather than bitwise operations. This should
+//!   preserve the pointer's provenance, which would otherwise be lost.
```

Comment:
Any reason not to use `wrapping_add`/`wrapping_sub`? Then we don't have to worry about going out-of-bounds...

Reply:
>  those methods were equivalent to casting through integers

No, that is not the case. (Where is that impression coming from? We might have to improve the docs.)
The docs currently say

> The resulting pointer “remembers” the allocated object that self points to; it must not be used to read or write other allocated objects.

which is pretty much saying that provenance is preserved. In fact, preserving the provenance is the entire reason those methods even exist!

Reply:
The existence of both of these implies there are cases where `ptr::{sub,add}` and `ptr::wrapping_{sub,add}` have different performance characteristics. Given that the non-wrapping nature of the use here is so straightforward, shouldn't we keep it this way?

Reply:
I found the in-bounds argument to be somewhat subtle, so using the unsafe arithmetic looks like a premature optimization to me. I.e., the difference in perf seems more theoretical than practical.

But, up to you, I am not at all an expert on the perf details here.

Reply:
Fair enough, I don't feel like looking at the asm for this (let alone doing benchmarks), so I've switched to `wrapping_{add,sub}`, but left the note that the `wrapping` isn't actually needed, in case someone needs to change it in the future.